File: code/WRSN-CEC2021-MMC-MFEA/src/algo/time/TIndividual.java
package algo.time;

import algo.path.PIndividual;
import elements.Charger;
import elements.Sensor;
import problem.Configs;
import problem.ProblemManager;

public class TIndividual {
	private static long counter = 0;
	private long id;

	private static int[][] path;
	private static double[][] greedyChargingTimeRatio;

	private double[] chargingTime;

	private double[] gene;
	private int skillFactor;
	private double scalarFitness;
	private double fitness;
	private int[] factorialRank;

	public TIndividual() {
		id = ++counter;
		gene = new double[ProblemManager.USSD];
		this.factorialRank = new int[ProblemManager.getTaskNumber()];
	}

	public void randomInit() {
		double sum = 0;
		for (int i = 0; i < gene.length; i++) {
			gene[i] = Configs.rand.nextDouble();
			sum += gene[i];
		}

		for (int i = 0; i < gene.length; i++) {
			gene[i] /= sum;
		}
	}

	public void initByGreedyTime() {
		for (int i=0; i<gene.length; i++) {
			if (i < greedyChargingTimeRatio[skillFactor].length) {
				gene[i] = greedyChargingTimeRatio[skillFactor][i];
			} else {
				gene[i] = Configs.rand.nextDouble();
			}
		}
	}

	public void decoding() {
		Charger ch = ProblemManager.chargers.get(skillFactor);
		double maxChargingTime = ch.getMaxChargingTime(path[skillFactor]);
		int n = path[skillFactor].length;
		this.chargingTime = new double[n];

		for (int i = 0; i < n; i++) {
			chargingTime[i] = maxChargingTime * gene[i];
		}
	}
	
	public void repairTime() {
		this.decoding();
		int n = path[skillFactor].length;
		Charger ch = ProblemManager.chargers.get(skillFactor);
		double maxChargingTime = ch.getMaxChargingTime(path[skillFactor]);
		
		
		double sum = 0;
		for (int i=0; i<n; i++) {
			if (chargingTime[i] < 0) {
				chargingTime[i] = 0;
			}
			sum += chargingTime[i];
		}
		
		if (sum > maxChargingTime) {
			double exceed = sum - maxChargingTime;
			
			double delta[] = new double[n];
			double sumDelta = 0;
			for (int i=0; i<n; i++) {
				delta[i] = Configs.rand.nextDouble();
				sumDelta += delta[i];
			}
			
			for (int i=0; i<n; i++) {
				chargingTime[i] -= exceed * delta[i] / sumDelta;
			}
		}

		double[] arriveTime = new double[n];
		arriveTime[0] = ProblemManager.distance[ProblemManager.serviceStation.getId()][path[skillFactor][0]]
				/ ch.getSpeed();
		for (int i = 0; i < n; i++) {
			Sensor s = ProblemManager.getSensorById(path[skillFactor][i]);

			double eRemain = s.getE0() - arriveTime[i] * s.getP();
			double ub = (Configs.S_EMAX - eRemain) / (ch.getU() - s.getP());

			if (eRemain < Configs.S_EMIN || chargingTime[i] < 0) {
				chargingTime[i] = 0;
			} else if (chargingTime[i] > ub) {
				chargingTime[i] = ub;
			}

			if (i < n - 1) {
				arriveTime[i + 1] = arriveTime[i] + chargingTime[i]
						+ ProblemManager.distance[path[skillFactor][i]][path[skillFactor][i + 1]] / ch.getSpeed();
			}
		}
	}

	public double calculateFitness() {
		this.decoding();
		int dead = 0;
		double totalLifetime = 0;
		double maxLifetime = 0;
		
		int n = path[skillFactor].length;
		double[] arriveTime = new double[n];
		Charger ch = ProblemManager.chargers.get(skillFactor);
		boolean visited[] = new boolean[ProblemManager.maxSensorId + 1];

		// repairing
		this.repairTime();

		// calculate fintess
		for (int i = 0; i < n; i++) {
			Sensor s = ProblemManager.getSensorById(path[skillFactor][i]);
			visited[s.getId()] = true;
			double eRemain = s.getE0() - arriveTime[i] * s.getP();
			double eAfterT = s.getE0() - Configs.T * s.getP() + chargingTime[i] * ch.getU();

			if (eRemain < s.getEmin() || eAfterT < s.getEmin()) {
				dead++;
			}
			
			double lifetime = (s.getE0() + chargingTime[i] * ch.getU() - s.getEmin()) / s.getP();
			totalLifetime += lifetime;
			maxLifetime = Math.max(lifetime, maxLifetime);
		}

		for (Sensor s : ProblemManager.subNet[this.skillFactor]) {
			if (!visited[s.getId()]) {
				double eAfterT = s.getE0() - Configs.T * s.getP();
				if (eAfterT < Configs.S_EMIN) {
					dead++;
				}
				
				double lifetime = s.getLifetime();
				totalLifetime += lifetime;
				maxLifetime = Math.max(lifetime, maxLifetime);
			}
		}

		double netSize = ProblemManager.subNet[this.skillFactor].size();
		double networkSurvivability = (netSize - dead) / netSize;

		return networkSurvivability;
	}

	/**
	 * polynormial mutation
	 */
	public void mutation() {
		double u;
		for (int i = 0; i < gene.length; i++) {
			u = Configs.rand.nextDouble(); // generate random u in [0, 1]
			if (u <= 0.5) {
				double deltaL = Math.pow(2 * u, 1.0 / (Configs.NM + 1)) - 1;
				double p = this.getGene(i) * (1.0 + deltaL);
				this.setGene(i, Math.max(0, Math.min(1, p)));
			} else {
				double deltaR = 1.0 - Math.pow(2 * (1 - u), 1.0 / (Configs.NM + 1));
				double p = this.getGene(i) + deltaR * (1 - this.getGene(i));
				this.setGene(i, Math.max(0, Math.min(1, p)));
			}
		}
	}

	public double[] getGene() {
		return gene;
	}

	public void setGene(double[] gene) {
		this.gene = gene;
	}

	public void setGene(int index, double value) {
		gene[index] = value;
	}

	public double getGene(int index) {
		return gene[index];
	}

	public int getSkillFactor() {
		return skillFactor;
	}

	public void setSkillFactor(int skillFactor) {
		this.skillFactor = skillFactor;
	}

	public double getScalarFitness() {
		return scalarFitness;
	}

	public void setScalarFitness(double scalarFitness) {
		this.scalarFitness = scalarFitness;
	}

	public double getFitness(int task) {
		if (task == this.skillFactor) {
			return fitness;
		} else {
			return -Double.MAX_VALUE;
		}
	}

	public void setFitness(double value) {
		fitness = value;
	}

	public int[] getFactorialRank() {
		return factorialRank;
	}

	public int getFactorialRank(int task) {
		return factorialRank[task];
	}

	public void setFactorialRank(int[] factorialRank) {
		this.factorialRank = factorialRank;
	}

	public void setFactorialRank(int task, int factorialRank) {
		this.factorialRank[task] = factorialRank;
	}

	public long getId() {
		return id;
	}

	public double[] getChargingTime() {
		return chargingTime;
	}

	public void setChargingTime(double[] chargingTime) {
		this.chargingTime = chargingTime;
	}

	public double getChargingTime(int index) {
		return this.chargingTime[index];
	}

	public PIndividual clone() {
		PIndividual indiv = new PIndividual();
		indiv.setGene(this.getGene().clone());
		return indiv;
	}

	public static int[][] getPath() {
		return path;
	}

	public static void setPath(int[][] path) {
		TIndividual.path = path;
	}

	public double getFitness() {
		return fitness;
	}

	public static void setGreedyChargingTime(double[][] time) {
		TIndividual.greedyChargingTimeRatio = new double[time.length][];

		for (int a = 0; a < time.length; a++) {
			double maxChargingTime = ProblemManager.chargers.get(a).getMaxChargingTime(path[a]);
			TIndividual.greedyChargingTimeRatio[a] = new double[time[a].length];
			for (int i = 0; i < time[a].length; i++) {
				TIndividual.greedyChargingTimeRatio[a][i] = time[a][i] / maxChargingTime;
			}
		}
	}
}

File: algo/time/TIndividual.py
import sys

from algo.path.PIndividual import PIndividual
from elements.Charger import Charger
from elements.Sensor import Sensor
from problem.Configs import Configs
from problem.ProblemManager import ProblemManager
from typing import overload

class TIndividual:
	#private static long counter = 0;
    counter = 0
    """private static long counter"""

	#private static int[][] path;
    path: 'list[list[int]]'
    """private static int[][] path"""
    path = []   

	#private static double[][] greedyChargingTimeRatio;
    greedyChargingTimeRatio: 'list[list[float]]'
    """private static double[][] greedyChargingTimeRatio"""
    greedyChargingTimeRatio = []

	#private long id;
    id: int

	#private double[] chargingTime;
    chargingTime: 'list[float]'
    chargingTime = []

    #private double[] gene;
    gene: 'list[float]'
    gene = []

    #private int skillFactor;
    skillFactor: int

    #private double scalarFitness;
    scalarFitness: float

    #private double fitness;
    fitness: float

    #private int[] factorialRank;
    factorialRank: 'list[int]'
    factorialRank = []

    def __init__(self) -> None:
        TIndividual.counter += 1
        self.id = self.counter
        self.gene = []
        for i in range(ProblemManager.USSD):
            self.gene.append(float())
        self.factorialRank = []
        for i in range(ProblemManager.getTaskNumber()):
            self.factorialRank.append(int())


    def randomInit(self) -> None:
        sum = 0.0
        for i in range(len(self.gene)):
            self.gene[i] = Configs.rand.uniform(0,1)
            sum += self.gene[i]

        for i in range(len(self.gene)):
            self.gene[i] /= sum


    def initByGreedyTime(self) -> None:
        for i in range(len(self.gene)):
            if (i < len(self.greedyChargingTimeRatio[self.skillFactor])):
                self.gene[i] = self.greedyChargingTimeRatio[self.skillFactor][i]
            else:
                self.gene[i] = Configs.rand.uniform(0,1)
            
    def decoding(self) -> None:
        ch = ProblemManager.chargers[self.skillFactor]
        maxChargingTime = ch.getMaxChargingTime(TIndividual.path[self.skillFactor])
        n = len(TIndividual.path[self.skillFactor])


        for i in range(n):
            self.chargingTime.append(maxChargingTime * self.gene[i]) 

    def repairTime(self) -> None:
        self.decoding()
        n = len(TIndividual.path[self.skillFactor])
        ch = ProblemManager.chargers[self.skillFactor]
        maxChargingTime = ch.getMaxChargingTime(TIndividual.path[self.skillFactor])
        

        sum = 0.0
        for i in range(n):
            if (self.chargingTime[i] < 0):
                self.chargingTime[i] = 0
            sum += self.chargingTime[i]
        
        if (sum > maxChargingTime):
            exceed = sum - maxChargingTime
            delta = []
            sumDelta = 0.0
            for i in range(n):
                delta.append(Configs.rand.uniform(0,1))
                sumDelta += delta[i]
            
            for i in range(n):
                self.chargingTime[i] -= exceed * delta[i] / sumDelta
            
        arriveTime: 'list[float]'
        arriveTime = [0.0] * n
        arriveTime[0] = ProblemManager.distance[ProblemManager.serviceStation.id][TIndividual.path[self.skillFactor][0]] / ch.speed
        for i in range(n):
            s = ProblemManager.getSensorById(TIndividual.path[self.skillFactor][i])

            eRemain = s.E0 - arriveTime[i] * s.p
            ub = (Configs.S_EMAX - eRemain) / (ch.U - s.p)

            if eRemain < Configs.S_EMIN or self.chargingTime[i] < 0:
                self.chargingTime[i] = 0
            elif self.chargingTime[i] > ub:
                self.chargingTime[i] = ub
            

            if i < n - 1:
                arriveTime[i + 1] = arriveTime[i] + self.chargingTime[i] \
                        + ProblemManager.distance[TIndividual.path[self.skillFactor][i]][TIndividual.path[self.skillFactor][i + 1]] / ch.speed

    def calculateFitness(self) -> float:
        self.decoding()
        dead = 0
        totalLifetime = 0.0
        maxLifetime = 0.0
        
        n = len(TIndividual.path[self.skillFactor])
        arriveTime = [0.0] * n
        ch = ProblemManager.chargers[self.skillFactor]
        visited = [False]*(ProblemManager.maxSensorId + 1)

        # repairing
        self.repairTime()

        # calculate fintess
        for i in range(n):
            s = ProblemManager.getSensorById(TIndividual.path[self.skillFactor][i])
            visited[s.id] = True
            eRemain = s.E0 - arriveTime[i] * s.p
            eAfterT = s.E0 - Configs.T * s.p + self.chargingTime[i] * ch.getU()

            if eRemain < s.getEmin() or eAfterT < s.getEmin():
                dead += 1
            
            
            lifetime = (s.getE0() + self.chargingTime[i] * ch.getU() - s.getEmin()) / s.getP()
            totalLifetime += lifetime
            maxLifetime = max(lifetime, maxLifetime)
        

        for s in ProblemManager.subNet[self.skillFactor]:
            if not visited[s.getId()]:
                eAfterT = s.getE0() - Configs.T * s.getP()
                if eAfterT < Configs.S_EMIN:
                    dead+=1
                
                lifetime = s.getLifetime()
                totalLifetime += lifetime
                maxLifetime = max(lifetime, maxLifetime)

        netSize = len(ProblemManager.subNet[self.skillFactor])
        networkSurvivability = (netSize - dead) / netSize

        return networkSurvivability


    
    def mutation(self) -> None:
        """polynormial mutation"""
        u = 0.0
        for i in range(len(self.gene)):
            u = Configs.rand.uniform(0,1); # generate random u in [0, 1]
            if (u <= 0.5):
                deltaL = pow(2 * u, 1.0 / (Configs.NM + 1)) - 1
                p = self.gene[i] * (1.0 + deltaL)
                self.gene[i] = max(0, min(1, p))
            else:
                deltaR = 1.0 - pow(2 * (1 - u), 1.0 / (Configs.NM + 1))
                p = self.gene[i] + deltaR * (1 - self.gene[i])
                self.gene[i] = max(0, min(1, p))



    def setGene(self, _gene: 'list[float]'):
        self.gene = _gene



    # public void setGene(int index, double value):
    #     gene[index] = value;
    # }



    # public int getSkillFactor() {
    #     return skillFactor;
    # }

    # public void setSkillFactor(int skillFactor) {
    #     this.skillFactor = skillFactor;
    # }

    # public double getScalarFitness() {
    #     return scalarFitness;
    # }

    # public void setScalarFitness(double scalarFitness) {
    #     this.scalarFitness = scalarFitness;
    # }

    def getFitness(self, task: int) -> float:
        if (task == self.skillFactor):
            return self.fitness
        else:
            return sys.float_info.max

    # public void setFitness(double value) {
    #     fitness = value;
    # }

    # public int[] getFactorialRank() {
    #     return factorialRank;
    # }

    # public int getFactorialRank(int task) {
    #     return factorialRank[task];
    # }

    # public void setFactorialRank(int[] factorialRank) {
    #     this.factorialRank = factorialRank;
    # }

    # public void setFactorialRank(int task, int factorialRank) {
    #     this.factorialRank[task] = factorialRank;
    # }

    # public long getId() {
    #     return id;
    # }

    # public double[] getChargingTime() {
    #     return chargingTime;
    # }

    # public void setChargingTime(double[] chargingTime) {
    #     this.chargingTime = chargingTime;
    # }

    # public double getChargingTime(int index) {
    #     return this.chargingTime[index];
    # }

    def clone(self) -> PIndividual:
        indiv =  PIndividual()
        indiv.gene = self.gene.copy()
        return indiv

    # public static int[][] getPath() {
    #     return path;
    # }

    # public static void setPath(int[][] path) {
    #     TIndividual.path = path;
    # }

    # public double getFitness() {
    #     return fitness;
    # }
    @staticmethod
    def setGreedyChargingTime(time: 'list[list[float]]') -> None:
        n = len(time)

        TIndividual.greedyChargingTimeRatio = [] * n

        for a in range(n):
            maxChargingTime = ProblemManager.chargers[a].getMaxChargingTime(TIndividual.path[a])
            TIndividual.greedyChargingTimeRatio[a] = [0.0] * len(time[a])
            for i in range(len(time[a])):
                TIndividual.greedyChargingTimeRatio[a][i] = time[a][i] / maxChargingTime


